### ISSUE TITLE
Dragging Files Between Windows

### ISSUE THREAD

**Issue by TediouslyBored** [jderose9/dash-to-panel#46]
Would it be possible to allow for files to be dragged between windows? It would be nice to be able to drag a file from one window, hover it over the icon or preview of a second window, and have the second window be brought to the front. This would be helpful for dragging files from a file manager to Google Drive.

**Comment by moriel5** [comment]
+1.
This personally is a pet peeve of mine, as I am used to doing that from back when I had used to be a Windows user.

**Comment by franglais125** [comment]
AFAIK:
This is not easy to implement within an extension. For more info you can take a lok at: https://github.com/micheleg/dash-to-dock/issues/49, where the discussion took place for that extension, and the corresponding gnome bug is listed, with no solution.

tldr: functionality has to be implemented upstream before it can be done at the extension level.

**Comment by TediouslyBored** [comment]
May not be entirely related but with dash to panel disabled, files can be dragged to the top left hot corner and dragged over another window. With dash to panel enabled, dragging the file to the hot corner does not show the activities overview. Can this be fixed, or is it intentional?

**Comment by moriel5** [comment]
Thanks, I had read @micheleg's answer, however I did not exactly understand the part about opening the windows.
Did he mean that it opens the activity view, or the corresponding window (which is what I need)?

**Comment by micheleg** [comment]
Basically the shell interface can't get information about which file is being dragged, but it can know a file is being dragged. But application windows can. My idea - never implemented - was on drag to show the open windows of the selected application, then hover the desired window to activate it, then drop. This is exactly the same dynamics you get by dragging a file to the activities button, and then look for the desired window. Showing only the windows of the selected application just makes it quicker.

**Comment by moriel5** [comment]
Thanks for the quick reply!
Yeah, I have heard of GNOME's developers' notoriety of ignoring perfectly sound recommendations that do not originate from their own team, however it would seem that your idea was at least considered, but afterwards completely ignored, it's a shame.

There are no workarounds, other than calling upon the window view? Perhaps some custom code from the extension itself (although that will unfortunately limit the possibilites of integration with other extensions, and/or be too much to handle), or a seperate extensions that could "patch" the shell to allow other extensions to make use of it?

**Comment by micheleg** [comment]
It is my understanding that you really have to patch the C layer, which is just not possible from an extension. I don't blame GNOME developers though. It is probably a lack of direct interest for the feature - after all they don't need that functionality for the shell itself. I honestly haven't tried to push for the patch to be applied upstream. You could  test and update the patch to current upstream, and try to push for merging.

**Comment by moriel5** [comment]
Thank you very much.
Unfortunately, I simply do not have the time to so currently, however as soon as I'll be able to do so, I'll conduct tests and try pushing for an upstream merge, as it would be a shame for this feature to not be integrated.

The extension ecosystem should not be underestimated (I am also an Xposed Framework user on Android (whenever possible))!

**Comment by jderose9** [comment]
I definitely find this annoying too. While dropping the file directly onto the icon would be best, I think even just raising all the application's running windows when hovering the icon while dragging a file would be a big improvement without a lot of effort.

**Comment by jderose9** [comment]
The best solution I could get working in a reasonable amount of time was to popup the overview when you drag a file over the panel. Then you can drag over the window or workspace you are interested in and hover over the new window. After a couple of seconds that window will be raised and you can drop your file into it.

I wasn't able to find out anything about the contents of the dragged item, as @micheleg mentioned.

My next thought was to allow the user to hover over an icon and open the live preview window, then hover over one of those windows to raise it. But, grabbing focus by the preview window failed since it was already grabbed by the drag-and-drop handler, and the windows underneath were picking up the events. I tried a bunch of workarounds but basically no events get through to shell extensions except the drag-and-drop events in that state. You only know that a drag started, the mouse moved, and a drag ended.

I'm sure that with the inclination, it would be possible to override the grab handler to add some special cases, but even then you can't actually drop the item there without the C shell fix in place.

**Comment by jderose9** [comment]
Re-opening as I haven't been happy with how cumbersome the current solution is. Hopefully someone can take initiative to push for the necessary upstream changes.

**Comment by brofjst** [comment]
+1
Since the last update i was able to drag files to another window by sliding the cursor on the panel, with the last update this feature does not work anymore.
OS: Arch
gnome-shell-extension-dash-to-panel-git 23.r54.g14ac72f-1
gnome-shell 1:3.34.0+152+g0fdbde910-1

I post [here](https://streamable.com/nxm45) a video just to be more clear (in this video i'm trying to drag a pick found on facebook into telegram).

**Comment by charlesg99** [comment]
Thanks for bringing that up @brofjst! Should now be fixed on master.

**Comment by brofjst** [comment]
Now it works, thank you!

**Comment by petrstepanov** [comment]
Hi, with respect to the drag and drop behavior. I noticed that when dash-to-dock autohide feature is enabled, and user tries to drag a file onto the dash-to-dock - the panel does not slide back on the screen.

**Comment by Alt37** [comment]
There's a [pull request](https://github.com/micheleg/dash-to-dock/pull/1416) for "Dash To Dock" which might help.

**Comment by BarbzYHOOL** [comment]
this is so problematic, the PR above looks pretty simple to add at least to dash to dock, duno for dash to panel

there was also this thing https://askubuntu.com/questions/1030623/how-to-drag-and-drop-files-to-open-application

**Comment by onlyreportingissues** [comment]
Very essential missing feature.

Couldn't this method be applied here also?: https://github.com/paperwm/PaperWM/pull/574

**Comment by francislavoie** [comment]
I think what I've implemented here will solve this issue? https://github.com/home-sweet-gnome/dash-to-panel/pull/2476

**Comment by Cingirakli1Dumbelek** [comment]
> Would it be possible to allow for files to be dragged between windows? It would be nice to be able to drag a file from one window, hover it over the icon or preview of a second window, and have the second window be brought to the front. This would be helpful for dragging files from a file manager to Google Drive.

it looks like it's fixed now (x11, gnome46) since it's working for me. But on https://github.com/home-sweet-gnome/dash-to-panel/issues/2218#issuecomment-2634581992 @charlesg99 said:

> Yes this feature used to work fine everywhere, then one day it broke on wayland, then later it broke on X11 and started to work again on wayland. Call it gnome-shell "improvements". There are a lot of issues opened on gnome repositories about drag and drop, most of them ignored. All I'm saying is that until mutter's dnd works correctly on both x11 and wayland, I don't see the point of fixing this to see it break again with the next gnome-shell version. I wish that wasn't the case though cause I like this feature.

Did anyone tested on GNOME 49-50?
